bgpd/bgp_mplsvpn.c

```diff
@@ -1336,6 +1336,231 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
+			   label, allocated, &blnc->nexthop);
+	if (allocated)
+		/* update the entry with the new label */
+		blnc->label = label;
+	else
+		/*
+		 * previously-allocated label is now invalid
+		 * eg: zebra deallocated the labels and notifies it
+		 */
+		blnc->label = MPLS_INVALID_LABEL;
+
+	if (old_label == blnc->label)
+		return 0; /* no change */
+
+	/* update paths */
+	if (blnc->label != MPLS_INVALID_LABEL)
+		bgp_zebra_send_nexthop_label(ZEBRA_MPLS_LABELS_ADD, blnc->label,
+					     ZEBRA_LSP_BGP, &blnc->nexthop);
+
+	LIST_FOREACH (pi, &(blnc->paths), label_nh_thread) {
+		if (!pi->net)
+			continue;
+		table = bgp_dest_table(pi->net);
+		if (!table)
+			continue;
+		vpn_leak_from_vrf_update(blnc->to_bgp, table->bgp, pi);
+	}
+
+	return 0;
+}
+
+/* get label for a given bgp update
+ * return the label value:
+ * - per-vrf value:
+ *      if nexthop of path is not valid or there are no nexthops
+ *      if nexthop of path is not connected
+ *      if path becomes ECMP
+ *      if blackhole is found
+ * - valid per-nexthop value
+ *      if nexthop of path is valid and connected and a label has been obtained
+ * - invalid value
+ *      if nexthop of path is valid and connected and label has not yet been
+ * obtained
+ */
+static mpls_label_t bgp_mplsvpn_get_path_info_label(struct bgp_path_info *pi,
+						    struct bgp *to_bgp,
+						    struct bgp *from_bgp,
+						    afi_t afi, safi_t safi)
+{
+	struct bgp_nexthop_cache *bnc;
+	struct bgp_label_per_nexthop_cache *blnc;
+	struct bgp_label_per_nexthop_cache_head *tree;
+	struct prefix *nh_to_use_ptr;
+	struct prefix nh_from_bnc = {0};
+	struct bgp *bgp_nexthop;
+	afi_t nh_afi;
+
+	/* check the next-hop reachability in the
+	 * bgp instance where the next-hop of the entry is
+	 */
+	if (pi->extra && pi->extra->bgp_orig)
+		bgp_nexthop = pi->extra->bgp_orig;
+	else
+		bgp_nexthop = from_bgp;
+
+	nh_afi = BGP_ATTR_NH_AFI(afi, pi->attr);
+
+	if (bgp_find_or_add_nexthop(from_bgp, bgp_nexthop, nh_afi, safi, pi,
+				    NULL, 0, NULL) == 0)
+		/* no valid next-hop, wait for next-hop validity */
+		return MPLS_INVALID_LABEL;
+
+	bnc = pi->nexthop;
+	if (!bnc) {
+		/* no valid next-hop, then per-vrf label is chosen
+		 * note that the case nexthop_num == 0 should not happen
+		 * as the 'bgp_process_nexthop_update' function frees the bnc
+		 * entry
+		 */
+		return from_bgp->vpn_policy[afi].tovpn_label;
+	}
+
+	/* check that the number of available nexthop is 1
+	 */
+	if (bnc->nexthop_num > 1 ||
+	    bnc->nexthop->type == NEXTHOP_TYPE_BLACKHOLE) {
+		/* ECMP and blachole routes do not support
+		 * per-nexthop label mode;
+		 * fallback to the per-vrf label.
+		 */
+		bgp_mplsvpn_path_nh_label_unlink(pi);
+		return from_bgp->vpn_policy[afi].tovpn_label;
+	}
+
+	/* extract the resolved nexthop given by the BNC nexthop cache
+	 * if nexthop type is not connected, then use the resolved nexthop
+	 */
+	if (bnc->nexthop->type != NEXTHOP_TYPE_IFINDEX) {
```

Comment:
Obviously, you need a `switch(bnc->nexthop->type)` and to move the case `NEXTHOP_TYPE_BLACKHOLE` here

Reply:
done